src/extract.rs

```diff
@@ -61,71 +68,114 @@ impl EGraph {
 }
 
 impl<'a> Extractor<'a> {
-    fn new(egraph: &'a EGraph) -> Self {
+    pub fn new(egraph: &'a EGraph, termdag: &mut TermDag) -> Self {
         let mut extractor = Extractor {
             costs: HashMap::default(),
             egraph,
             ctors: vec![],
         };
 
-        // HACK
-        // just consider all functions constructors for now...
-        extractor.ctors.extend(egraph.functions.keys().cloned());
+        // only consider "extractable" functions
+        extractor.ctors.extend(
+            egraph
+                .functions
+                .keys()
+                .filter(|func| !egraph.functions.get(*func).unwrap().decl.unextractable)
+                .cloned(),
+        );
 
         log::debug!("Extracting from ctors: {:?}", extractor.ctors);
-        extractor.find_costs();
+        extractor.find_costs(termdag);
         extractor
     }
 
-    fn expr_from_node(&self, node: &Node) -> Expr {
-        let children = node.inputs.iter().map(|&value| {
-            let arcsort = self.egraph.get_sort(&value).unwrap();
-            self.find_best(value, arcsort).1
-        });
-        Expr::call(node.sym, children)
+    fn expr_from_node(&self, node: &Node, termdag: &mut TermDag) -> Term {
+        let mut children = vec![];
+        for value in node.inputs {
+            let arcsort = self.egraph.get_sort(value).unwrap();
+            children.push(self.find_best(*value, termdag, arcsort).1)
+        }
+
+        termdag.make(node.sym, children)
     }
 
-    fn find_best(&self, value: Value, sort: &ArcSort) -> (Cost, Expr) {
+    pub fn find_best(&self, value: Value, termdag: &mut TermDag, sort: &ArcSort) -> (Cost, Term) {
         if sort.is_eq_sort() {
-            let id = self.egraph.find(Id::from(value.bits as usize));
-            let (cost, node) = &self
+            let id = self.find(&value);
+            let (cost, node) = self
                 .costs
                 .get(&id)
-                .unwrap_or_else(|| panic!("No cost for {:?}", value));
-            (*cost, self.expr_from_node(node))
+                .unwrap_or_else(|| {
+                    eprintln!("No cost for {:?}", value);
+                    for func in self.egraph.functions.values() {
+                        for (inputs, output) in func.nodes.iter() {
+                            if output.value == value {
+                                eprintln!("Found unextractable function: {:?}", func.decl.name);
```

Comment:
I think `eprintln!`s would be better off being replaced as log statements. These can probably be `log::error!`s. The nice thing about logs is that callers can opt out, which is useful when using egglog as a library.

Reply:
good idea